ssl/ssl_key_share.cc

```diff
@@ -176,20 +180,557 @@ class X25519KeyShare : public SSLKeyShare {
+      // Ensure |out| is valid, has no children, and has been initialized.
+      // We check that |out| has no children because otherwise CBB_data()
+      // will produce a fatal error by way of assert(cbb->child == NULL);
+      if (!out || out->child || !CBB_data(out)) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      if (!hybrid_group_ || this->exchange_performed) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // Iterate through the component groups and Offer() each of their key
+      // shares. If one of the calls to a component Offer() fails,
+      // OPENSSL_PUT_ERROR will be set appropriately in that component.
+      for (const UniquePtr<SSLKeyShare> &key_share : key_shares_) {
+        if (!key_share || !key_share->Offer(out)) {
+          return false;
+        }
+      }
+      this->exchange_performed = true;
+      return true;
+    }
+
+    bool Accept(CBB *out_public_key, Array<uint8_t> *out_secret,
+                uint8_t *out_alert, Span<const uint8_t> peer_key) override {
+      if (!out_alert) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      // Set alert to internal error by default
+      *out_alert = SSL_AD_INTERNAL_ERROR;
+
+      if (!out_secret|| !peer_key.data()) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      // Ensure |out_public_key| is valid, has no children, and has been
+      // initialized. We check that |out_public_key| has no children because
+      // otherwise CBB_data() will produce a fatal error by way of
+      // assert(cbb->child == NULL);
+      if (!out_public_key || out_public_key->child || !CBB_data(out_public_key)) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return false;
+      }
+
+      if (!hybrid_group_ || this->exchange_performed) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // A hybrid shared secret with two components should be 64 bytes.
+      // If it happens to be larger, the CBB will grow accordingly.
+      CBB hybrid_shared_secret;
+      if (!CBB_init(&hybrid_shared_secret, 64)) {
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // Accept() each component key share. Each component's Accept() function
+      // will generate a shared secret and a public key to be sent back to
+      // the peer. The hybrid public key is the concatenation of all component
+      // public keys; the hybrid shared secret is the concatenation of all
+      // component shared secrets.
+      size_t peer_key_index = 0;
+      for (size_t i = 0; i < NUM_HYBRID_COMPONENTS; i++) {
+        size_t component_key_size = 0;
+        if (!get_component_offer_key_share_size(&component_key_size, hybrid_group_->component_group_ids[i])) {
```

Comment:
What logic should be moved to the constructor? Do you mean that we should use an array lookup instead of a getter function, and add and initialize two new private class variables `uint16_t offer_share_size[2]` and `uint16_t accept_share_size[2]` to the HybridKeyShare class?

Reply:
yes, I meant something like that. Or maybe even have the base `SSLKeyShare` class have fields for the share sizes? Seems to me like a normal information to have in this class, but I'm not very familiar with this code so I might be wrong :)